### ISSUE TITLE
Automatic scaling of Image.MAXBLOCK

### ISSUE THREAD

**Issue by acdha** [python-pillow/Pillow#148]
There's a venerable old PIL bug which caused `Image.save(optimize=True)` to fail with JPEGs when the file size is over a certain threshold. The earliest reference I could find dates back to 1999:

http://mail.python.org/pipermail/image-sig/1999-December/000942.html

The workaround commonly used is to either hardcode something like this:

```
ImageFile.MAXBLOCK = <empirically determined constant (i.e. wild guess)>
```

or the more dynamic approach used in e.g. jdriscoll/django-imagekit#50:

```
ImageFile.MAXBLOCK = max(img.size) ** 2
```

This seems like exactly the kind of thing Pillow could just fix once and for all.

**Comment by agschwender** [comment]
One of the users of Pilbox reported that they were receiving an `encoder error -2 when writing image file` error. I did some digging into it and it turned out it was receiving a `Suspension not allowed here`. Looking at the code it seemed as if it might be caused by a combination of `quality=keep` and `optimize=True`.  Specifically, it looked as if it would not reach the condition that increased the buffer size. 

```
if "optimize" in info or "progressive" in info or "progression" in info:
    if quality >= 95:
        bufsize = 2 * im.size[0] * im.size[1]
    else:
        bufsize = im.size[0] * im.size[1]
```

Here's a reference to the issue: https://github.com/agschwender/pilbox/issues/29

**Comment by Apreche** [comment]
I know this bug is marked as closed, but I have come across a jpg file uploaded by a user that triggered this error. I am using Python 2.7.6 and Pillow 2.8.1. The error also exists in 2.8.0 and 2.7.0. 

Here are steps to reproduce.

1. Open the JPEG file
2. Extract the icc_profile data
3. Thumbnail the jpeg, but don't make the thumbnail much smaller than the original.
4. Save it as a jpeg with optimize=True, the original icc_profile, and quality=95

Expected Result: Save a thumbnailed image
Actual Result: Prints the message "Suspension not allowed here" and raises an "IOError: encoder error -2" 

Here is some code to reproduce the error.

```Python
from PIL import Image
from StringIO import StringIO

i = Image.open('bad.jpg')
icc_profile = i.info.get('icc_profile')
i.thumbnail((127, 275), Image.ANTIALIAS)
output = StringIO()
i.save(output, format='JPEG', quality=95, optimize=True, icc_profile=icc_profile)
```

Here is the ![offending jpeg](https://cloud.githubusercontent.com/assets/5847/7435686/d168ee6a-f013-11e4-8293-a872718d1a37.jpg)

I think the problem is that this jpeg has a color depth of 32 bits. I did not think that was possible for a jpeg. When I re-encoded the image to be 24-bits using an image editing application, this error disappeared. The file itself clearly has a lot of problems.

Also, if the thumbnail I am making is very near to the original size of the image, the error occurs. If I make the thumbnail much smaller than the original image, the error will go away. The error can also go away if we set optimize=False, or decrease the quality setting.

I know that it is easy to just blame the problem on the bad image file and not fix this problem. However, despite the image being bad, every image editing and viewing program that I tried was able to handle this image properly. Imgur had no problems with it. Pillow even handles it properly in most cases. Certain combinations of parameters just cause it to fail.

**Comment by jaddison** [comment]
I've just had to downgrade to Pillow==2.7.0 to get image manipulation working with sorl-thumbnail again. I don't think we're doing anything particularly fancy with sorl.

**Comment by aclark4life** [comment]
@jaddison Can you elaborate on that please? Thanks

**Comment by jaddison** [comment]
I'm not sure how to explain it any better. With 2.8.1, I was getting quite a few "Suspension not allowed here" log messages in my supervisor.log file, along with a quite a few images that weren't getting thumbnailed.  Downgrading to 2.7.0 (I'd only upgraded yesterday, and that's when the problems started) resolved the problems.

Well... most of the problems. With 2.7.0, I'm having troubles with a much smaller number of images. This one is apparently one such pain:

![community-day-186239-ca25d3](https://cloud.githubusercontent.com/assets/101148/7526767/9d0f64a4-f4c9-11e4-8d0d-e29c60838f5d.jpg)

Let me know if I can run some tests for you.

**Comment by aclark4life** [comment]
Thanks @jaddison. Just trying to document in hopes we can find issue. I hate hearing about regressions :disappointed:

**Comment by aclark4life** [comment]
Since there have been so many comments, are we still talking about ``Automatic scaling of Image.MAXBLOCK`` or do we need a new issue?

**Comment by jaddison** [comment]
I have no idea! Tell me how to find out for you? :)

**Comment by wiredfool** [comment]
Set `ImageFile.MAXBLOCK` to something large and bigger than the filesize of any of your thumbnails, like 1M. Then run your tests.

**Comment by paregorios** [comment]
FWIW I'm seeing a very similar scenario to what @Apreche describes above (open, get icc_profile, save thumbnail with profile and optimize). Python 2.7.8. Pillow 2.8.1 (also saw it in 2.8.0).

**Comment by Apreche** [comment]
Yes, I can confirm that I'm seeing similar results as @paregorios.
 
I don't know much about how the internals of PIL(low) works, but is the MAXBLOCK related to space allocated for the entire image file, or just the image data portion? Could it be that although MAXBLOCK is big enough for the image data, the additional meta-data of the icc_profile puts it over the top? Perhaps adding the size of the icc_profile, and any other metadata like EXIF, to the MAXBLOCK will result in it being large enough?

**Comment by paregorios** [comment]
@Apreche It's my impression that pilkit's fix assumes that MAXBLOCK is related to the entire size of the image file, otherwise it wouldn't try to accommodate the size of the exif header. I see no attempt to address icc_profile size there, but I tried forcing it to in a local copy (with a guess at the prefix length), but got no joy. It's possible that something else involving image size and icc profile is making the old code dragons in the basement grumpy.

**Comment by wiredfool** [comment]
MAXBLOCK is the size of the buffer that we're passing to libjpeg.  Generally, libjpeg does things in a chunked manner. Exif info, other metadata, compression of pixels, all of it can be done in successive calls into the library. But there are certain things that have to be emitted in one shot, like each chunk of metadata (which is why the exif size shows up), each line of the image, and in some cases, the whole compressed image. There may be others, as this thread is showing.

**Comment by Darou** [comment]
I was having a similar problem with the bufsize and found that with some JPEG images,
the metadata in icc_profiles is larger than the MAXBLOCK and the image size.

That's due JpegImagePluging writting all the icc_profile data as a chunk, instead
of by markers.

The problem affects both images posted in May 2015 on this thread.

To test the problem you can use this test:

```
def test_large_icc_meta(self):
    # https://github.com/python-pillow/Pillow/issues/148
    # Sometimes the meta data on the icc_profile block is biger than
    # Image.MAXBLOCK and the image size.
    im = Image.open('Tests/images/icc_profile_big.jpg')
    f = self.tempfile("temp.jpg")
    icc_profile = im.info["icc_profile"]
    try:
        im.save(f, format='JPEG', progressive=True, quality=95,
                icc_profile=icc_profile, optimize=True)
        assert True

    except IOError:
        assert False, "Failed saving image with icc larger than image size"
```

To solve it we just have to add a check on the size of the icc_profile block:

```
bufsize = max(ImageFile.MAXBLOCK, bufsize, len(info.get("exif", b"")) + 5, len(extra) + 1)
```

**Comment by aclark4life** [comment]
@acdha @wiredfool I see some PRs, but I can't tell at a glance if this is fixed?

**Comment by wiredfool** [comment]
I thought it was

**Comment by acdha** [comment]
I believe so, too. It’s been a long time since I’ve needed to think about it.

**Comment by string-areeb** [comment]
Still fails for mozjpeg. Have to set `ImageFile.MAXBLOCK = 3 * image.size[0] * image.size[1]` before saving

https://blog.avirtualhome.com/memory-problems-with-jpg-files-and-pillow/

**Comment by iamareebjamal** [comment]
Still fails for mozjpeg. Have to set `ImageFile.MAXBLOCK = 3 * image.size[0] * image.size[1]` before saving

https://blog.avirtualhome.com/memory-problems-with-jpg-files-and-pillow/

**Comment by stefanistrate** [comment]
Hi everyone,

I don't think this has been fixed completely. Here's an image and a test for showing how it's currently crashing on an image with large EXIF data: https://github.com/stefanistrate/Pillow/commit/ab74c67e850eeabfdaf5392f251f311ce6381a69#diff-da0bdfd38e5a8aff97dc143bb3baba60

Image originally posted at: https://github.com/wagtail/Willow/pull/63#issuecomment-429050840.

**Comment by kgriffs** [comment]
I have also reproduced this bug when attempting to save a small image with a large EXIF payload:

Image Width: 300
Image Height: 225
Exif size: 12849

**Comment by alazakoa** [comment]
Can confirm this bug exists still as of Pillow 7.2.0, the code hasn't changed in this location in 8.0.0 so it's probably present too. I have a jpeg image in wagtail that's crashing my server. I can confirm it only happens for this specific jpeg right now, it's refocused before being passed off to pillow. The other image renditions using this jpeg are generated just fine, specifically width 1440, 720, and 165x165.

Occurs in Pillow version 5.4.1 and 7.2.0. I downgraded to see if the bug was still present when the project was initiated.

`Suspension not allowed` followed by:
```
app_1          |   File "/src/app/base/templatetags/srcset_tags.py", line 31, in get_srcset
app_1          |     image_sm = image.get_rendition("width-360")
app_1          |   File "/python_packages/lib/python3.6/site-packages/wagtail/images/models.py", line 304, in get_rendition
app_1          |     generated_image = filter.run(self, BytesIO())
app_1          |   File "/python_packages/lib/python3.6/site-packages/wagtail/images/models.py", line 454, in run
app_1          |     return willow.save_as_jpeg(output, quality=quality, progressive=True, optimize=True)
app_1          |   File "/python_packages/src/willow/willow/plugins/pillow.py", line 151, in save_as_jpeg
app_1          |     image.save(f, 'JPEG', quality=quality, **kwargs)
app_1          |   File "/python_packages/lib/python3.6/site-packages/PIL/Image.py", line 1994, in save
app_1          |     save_handler(self, fp, filename)
app_1          |   File "/python_packages/lib/python3.6/site-packages/PIL/JpegImagePlugin.py", line 761, in _save
app_1          |     ImageFile._save(im, fp, [("jpeg", (0, 0)+im.size, 0, rawmode)], bufsize)
app_1          |   File "/python_packages/lib/python3.6/site-packages/PIL/ImageFile.py", line 506, in _save
app_1          |     raise IOError("encoder error %d when writing image file" % s)
app_1          | OSError: encoder error -2 when writing image file
```

Would something like this work? I'm not sure of the implications but it would be my first guess. This seems to be a very odd heuristic rule.
```
 bufsize = max(MAXBLOCK, bufsize, im.size[0] * 4, im.size[1] * 4) 
```

**Comment by radarhere** [comment]
I'm unable to replicate the problems from @stefanistrate or @kgriffs.

@alazakoa could you open a new issue about this, and provide an example image?

**Comment by alazakoa** [comment]
@radarhere my example image only works in the context of Wagtail. I don't have a minimal example using only Pillow. 

I tired dumping the object with pickle and calling the operations again from a script but that didn't reproduce it, neither locally or in a container. I'm using debian right now with libjpeg62-turbo.

**Comment by radarhere** [comment]
This has been continued as #5448

**Comment by radarhere** [comment]
#5448 has now been resolved by #7345
